### DISCUSSION TITLE
Adding GPIO to any power monitoring device (Eg POWR316D)

### DISCUSSION THREAD

**Discussion by nitinuae** [arendst/Tasmota#16686] [Q&A]
Hello
As per a lot of discussions, all power monitoring devices have earth connected to ground. 
Is the above assumption correct? This removes the possibilty of adding any additional GPIO to the POWR316D that I wanted to have (I wanted to add pulse counter and one analog voltage input to it).

Curiously - How does POWR316TH devices add temperature reading?
Same with POWr2 (the old ones- esp8266 ones)?

I will be happy if someone can find me a solution.

**Comment by sfromis** [comment]
The general rule with energy monitoring devices is to **never** connect anything more, unless you have very specific knowledge of the "low voltage" side not being mains-connected. While I have not verified how the "Elite" devices are built, I did (with care) probe a Sonoff Pow R2, and found that the warning against connecting stuff was very well-founded, as this would be dangerous.

**Comment by barbudor** [comment]
There is no Sonoff POWR316TH 
There is POWR316 that do measure power consumption but do not have a connector for sensor
There are THR316 and THR320 that have a connector for sensor but no power measurment

There is roughly 2 ways of measuring power consumption:
- IN-LINE : by measuring the current going through a small resistor by measuring hte voltage across the pins of the resistor. To do that the measurement is in contact with the resistor which in direct contact with the mains lines. So the whole circuit, including the ESP is somehow connected to mains. Connecting something outside the box == DANGER.
This is the mode used by most home devices whereare their are plugs or relay boxes like all Sonoff , Shelly 1PM, etc...
- INDUCTION : The current through the mains cable is measured by a winding (a coil) around the mains line without direct contact. It works as a current transformer and the measuring circuit and ESP are fully isolated from the mains. Instead of a coil it can also use a Hall sensor which is a sensor sensitive to the magnetic field created by the wire.
This is what is used in PZEM, Shelly 3EM. Generally those devices are only measuring devices, without relays.

IN-LINE systems are cheapers and generally more accurate for low currnets than induction based but also most of the time limited in current (16A or rarely 20A). Anything above 20A would use induction measurment. 
It is not impossible to do In-line measurment above 20A, just much more risky.
It is not impossible to do isolated in-line measurment by isolating hte measurment circuit from the MCU/ESP but that is more expensive than what we are ready to afford for those home devices (requires dual power supplies, opto isolators)

So yes there are very good reasons for not connecting any sensor to a power measurment device.

**Comment by lalo-uy** [comment]
Ther are also Hall Current sensor, like ASC770, that keep the hi voltage
isolated. Don’t know about $$$

**Reply by nitinuae** [reply]
I have a few CTs lying around. I am thinking to use them (in ADC port of ESP32)

**Reply by sfromis** [reply]
Sure, there exists technology to measure power with galvanic isolation, but I'd just not expect such in the "standard" realm of cheap IoT energy monitoring devices, as the current method is both cheap and works well. Just very unsuitable for tinkerers with the mistaken belief that they have a platform to build on.

**Reply by barbudor** [reply]
@lalo-uy , You may have missed that sentence in my answer 😄 :

> Instead of a coil it can also use a *Hall sensor* which is a sensor sensitive to the magnetic field created by the wire.

**Comment by lalo-uy** [comment]
In deed, sorry
